content/doc/book/pipeline/pipeline-best-practices.adoc

```diff
@@ -0,0 +1,108 @@
+---
+layout: section
+title: Pipeline Best Practices
+---
+ifdef::backend-html5[]
+:notitle:
+:description:
+:author: Alex Taylor
+:email: jenkinsci-docs@googlegroups.com
+:sectanchors:
+ifdef::env-github[:imagesdir: ../resources]
+ifndef::env-github[:imagesdir: ../../resources]
+:hide-uri-scheme:
+:toc:
+endif::[]
+
+= Applying best practices to Pipelines
+
+This guide provides a small selection of best practices for pipelines and points out the most common mistakes. 
+
+The goal is to point pipeline authors and maintainers towards patterns that result in better Pipeline execution and away from pitfalls they might otherwise not be aware of.
+This guide is not meant to be an exhaustive list of all possible Pipeline best practices but instead to provide a number of specific examples useful in tracking down common practices.
+Use it as a "do this" generally and not as an incredibly detailed "how-to".
+
+This guide is arranged by area, guideline, then listing specific examples. 
+
+== General
+
+=== Making sure to use Groovy code in Pipelines as glue
+
+Use Groovy code to connect a set of actions rather than as the main functionality of your Pipeline.
+In other words, instead of relying on Pipeline functionality (Groovy or Pipeline steps) to drive the build process forward, use single steps (such as `sh`) to accomplish multiple parts of the build. 
+Pipelines, as their complexity increases (the amount of Groovy code, number of steps used, etc.), require more resources (CPU, memory, storage) on the master.
+Think of Pipeline as a tool to accomplish a build rather than the core of a build.
+
+Example: Using a single Maven build step to drive the build through its build/test/deploy process.
+
+=== Avoiding complex Groovy code in Pipelines
+
+For a Pipeline, Groovy code *always* executes on master which means using master resources(memory and CPU). 
+Therefore, it is critically important to reduce the amount of Groovy code executed by Pipelines (this includes any methods called on classes imported in Pipelines).
+The following are the most common example Groovy methods to avoid using:
+. *JsonSlurper:* This function (and some other similar ones like XmlSlurper or readFile) can be used to read from a file on disk, parse the data from that file into a JSON object, and inject that object into a Pipeline using a command like JsonSlurper().parseText(readFile("$LOCAL_FILE")). This command loads the local file into memory on the master twice and, if the file is very large or the command is executed frequently, will require a lot of memory.
+.. Solution: Instead of using JsonSlurper, use a shell step and return the standard out. This shell would look something like this: `def JsonReturn = sh label: '', returnStdout: true, script: 'echo "$LOCAL_FILE"| jq "$PARSING_QUERY"'`. This will use agent resources to read the file and the $PARSING_QUERY will help pare down the file into a smaller size.
```

Comment:
did you mean parse?
```suggestion
.. Solution: Instead of using JsonSlurper, use a shell step and return the standard out. This shell would look something like this: `def JsonReturn = sh label: '', returnStdout: true, script: 'echo "$LOCAL_FILE"| jq "$PARSING_QUERY"'`. This will use agent resources to read the file and the $PARSING_QUERY will help parse down the file into a smaller size.
```